### PR TITLE
Pre-calculate total needed dai and vat.move only once

### PR DESCRIPTION
**Pull request by oldchili**

### CODE REVIEW

**Review by telome — APPROVED** [review]

File: src/FlapperUniV2.sol [hunk 1]
```diff
@@ -138,6 +138,18 @@ contract FlapperUniV2 {
         (reserveDai, reserveGem) = daiFirst ? (_reserveA, _reserveB) : (_reserveB, _reserveA);
     }
 
+
+    // The Uniswap invariant needs to hold before and after the swap.
```
Comment (sunbreak1211) [inline src/FlapperUniV2.sol:142]:
```suggestion
    // The Uniswap invariant needs to hold before and after the swap.
```

File: src/FlapperUniV2.sol [hunk 2]
```diff
@@ -138,6 +138,17 @@ contract FlapperUniV2 {
...
+    // Additionally, The deposited funds need to be in the same ratio as the reserves after the swap.
+    //
+    // (1)   reserveDai * reserveGem = (reserveDai + lot * 997 / 1000) * (reserveGem - bought)
+    // (2)   (total - lot) / bought =  (reserveDai + lot) / (reserveGem - bought)
+    //
+    // The solution for the these equations for variable `total` and `bought` is used below.
+    function _getTotalDai(uint256 wlot, uint256 reserveDai) internal pure returns (uint256 total) {
+        total = wlot * (997 * wlot + 1997 * reserveDai) / (1000 * reserveDai);
+    }
```
Comment (sunbreak1211) [inline src/FlapperUniV2.sol:150]:
```suggestion
    // The deposited funds need to be in the same ratio as the reserves after the swap:
    // (total - wlot) / buy =  (reserveDai + wlot) / (reserveGem - buy)
    //
    // The buy amount comes from the _getAmountOut function logic:
    // buy = wlot * 997 * reserveGem / (reserveDai * 1000 + wlot * 997)
    //
    // The solution for the these equations for variable `total` is used below:
    // https://www.wolframalpha.com/input?i2d=true&i=%5C%2840%29Divide%5BT+-+L%2CB%5D%5C%2841%29%3DDivide%5BD%2BL%2CG+-+B%5D+%5C%2844%29+B%3DDivide%5BL*997*G%2CD*1000%2BL*997%5D+%5C%2844%29+L%3E0+%5C%2844%29+D%3E0+%5C%2844%29+G%3E0+solve+for+T
    function _getTotalDai(uint256 wlot, uint256 reserveDai) internal pure returns (uint256 total) {
        total = wlot * (997 * wlot + 1997 * reserveDai) / (1000 * reserveDai);
    }
```

File: src/FlapperUniV2.sol [hunk 3]
```diff
@@ -133,9 +135,17 @@ contract FlapperUniV2 {
         emit File(what, data);
     }
 
-    function _getReserves() internal view returns (uint256 reserveDai, uint256 reserveGem) {
+    function _getReserves() internal returns (uint256 reserveDai, uint256 reserveGem) {
         (uint256 _reserveA, uint256 _reserveB,) = pair.getReserves();
         (reserveDai, reserveGem) = daiFirst ? (_reserveA, _reserveB) : (_reserveB, _reserveA);
+
+        uint256 _daiDiff = GemLike(dai).balanceOf(address(pair)) - reserveDai;
```
Comment (telome) [inline src/FlapperUniV2.sol:142]: Small nit: maybe save the balances in `_daiBal`, `_gemBal` and use `if(_daiBal > reserveDai || _gemBal > reserveGem) { pair.sync(); reserveDai = _daiBal; reserveGem = _gemDiff; }` to save a a bit on arithmetic ops?

Comment (oldchili) [reply src/FlapperUniV2.sol]: Nice. Done.

File: src/FlapperUniV2.sol [hunk 4]
```diff
@@ -36,6 +36,7 @@ interface GemLike {
     function decimals() external view returns (uint8);
     function approve(address, uint256) external;
     function transfer(address, uint256) external;
+    function balanceOf(address) external view returns (uint256);
```
Comment (sunbreak1211) [inline src/FlapperUniV2.sol:39]: nit: I'd put it above to group with the other `view` as in the rest of interfaces.

File: src/FlapperUniV2.sol [hunk 5]
```diff
@@ -133,9 +135,28 @@ contract FlapperUniV2 {
...
+            reserveDai = _daiBalance;
+            reserveGem = _gemBalance;
+        }
+    }
+
+    // The Uniswap invariant needs to hold through the swap.
+    // Additionally, The deposited funds need to be in the same ratio as the reserves after the swap.
+    //
+    // (1)   reserveDai * reserveGem = (reserveDai + lot * 997 / 1000) * (reserveGem - bought)
+    // (2)   (total - lot) / bought =  (reserveDai + lot) / (reserveGem - bought)
```
Comment (sunbreak1211) [inline src/FlapperUniV2.sol:155]:
```suggestion
    // (2)   (total - lot) / bought  =  (reserveDai + lot) / (reserveGem - bought)
```

File: src/FlapperUniV2.sol (hunk 4 above)
Comment (oldchili) [reply src/FlapperUniV2.sol]: Done.

File: src/FlapperUniV2.sol [hunk 6]
```diff
@@ -150,14 +171,17 @@ contract FlapperUniV2 {
...
-        vat.move(msg.sender, address(this), lot);
-        daiJoin.exit(address(this), _wlot);
+        vat.move(msg.sender, address(this), _total * RAY);
+        daiJoin.exit(address(this), _total);
+        //
 
-        (uint256 _reserveDai, uint256 _reserveGem) = _getReserves();
+        // Swap
         uint256 _buy = _getAmountOut(_wlot, _reserveDai, _reserveGem);
         require(_buy >= _wlot * want / (uint256(pip.read()) * RAY / spotter.par()), "FlapperUniV2/insufficient-buy-amount");
```
Comment (sunbreak1211) [inline src/FlapperUniV2.sol:186]: Nit: we might want to do this before the `vat.move`. It seems like better to solve all the requires before doing the actual transfer (thinking that in this case is possible).

Comment (oldchili) [reply src/FlapperUniV2.sol]: Done.

**Pull request merged**